src/rfc/detail/rules.cpp

```diff
@@ -397,6 +398,75 @@ remove_obs_fold(
     }
 }
 
+//------------------------------------------------
+
+system::result<hex_rule_t::value_type>
+hex_rule_t::
+parse(
+    char const*& it,
+    char const* end) const noexcept
+{
+    value_type v;
+
+    constexpr
+    auto const max =
+        std::numeric_limits<std::size_t>::max();
+
+    constexpr
+    std::size_t const m1 = (max >> 4);
+
+    auto begin = it;
+
+    while( it < end )
+    {
+        auto n = grammar::hexdig_value(*it);
+        if(n < 0)
+        {
+            if( it == begin )
+                BOOST_HTTP_PROTO_RETURN_EC(
+                    grammar::error::mismatch);
+
+            return v;
+        }
+
+        if( v.v > m1 )
+            BOOST_HTTP_PROTO_RETURN_EC(
+                grammar::error::invalid);
+
+        if( v.v > max - n )
+            BOOST_HTTP_PROTO_RETURN_EC(
+                grammar::error::invalid);
+
+        v.v = (v.v << 4) + n;
```

Comment:
The check for overflow can be done with one less `if` statement. Just check whether the new value of `v.v` is less than or equal to the old value. If yes, then overflow has occurred. This is well defined as long as `v.v` is an unsigned integer.

Reply:
Unfortunately, I think there's a bug in my checks here anyway.
Maybe we should add a formal test here.
But for compound statements, a single if-statement won't suffice.
We need something like this: https://godbolt.org/z/Px8hbhToa

Each part of the expression needs to be individually checked for overflow, it seems.

Reply:
Upon further testing, my `<<` check was also wrong. What I have now is the ultimately correct thing to do in terms of checking for potential overflow. The only difference we could make is how we check for overflow when adding.

Reply:
I believe the proper check is:

```
if( v.v > (m - n) >> 4 )
  return "overflow";
```